File: src/features/sea/query_plan.zig
//! QueryPlan / TaskType inference for SEA retrieval.
//!
//! A free-text query is turned into a small, structured plan of tuning knobs by
//! a *deterministic keyword heuristic* — no model call — mirroring the SEA
//! reference design (`docs/spec/sea-design-extract.md` §4). The plan's flags
//! (notably `exact_recall`) let `gatherEvidence` make retrieval task-aware.

const std = @import("std");

/// Recognized retrieval intents. Inferred from the query by `infer`.
pub const TaskType = enum(u8) {
    general,
    implementation_design,
    code_repair,
    legal_review,
    research_synthesis,
    project_recall,
    benchmark_review,

    pub fn text(self: TaskType) []const u8 {
        return @tagName(self);
    }
};

/// Inferred, structured intent for a query. `query` is borrowed (never owned by
/// the plan). The scalars are reserved tuning knobs with sane defaults; SEA's
/// Phase-1 retrieval consumes `exact_recall` today (and `recency_bias` is
/// carried for future task-aware scoring).
pub const QueryPlan = struct {
    task: TaskType = .general,
    query: []const u8,
    /// Answers should be backed by selected evidence (design intent flag).
    require_grounding: bool = true,
    /// Trust provenance + exact wording over fuzzy semantic similarity. Set by
    /// `project_recall`; re-weights `gatherEvidence` toward lexical overlap.
    exact_recall: bool = false,
    /// Relative preference for fresher records, `[0,1]`. Reserved tuning knob.
    recency_bias: f32 = 0.40,
    /// Risk posture of the task, `[0,1]`. Reserved tuning knob.
    risk: f32 = 0.50,
};

/// One keyword group: any substring match maps the query to `task`.
const KeywordGroup = struct {
    task: TaskType,
    keywords: []const []const u8,
};

/// Keyword groups in ascending precedence — later groups overwrite earlier
/// matches (last-match-wins, per the reference design). `project_recall` is last
/// so an explicit recall intent dominates and enables `exact_recall`.
const KEYWORD_GROUPS = [_]KeywordGroup{
    .{ .task = .implementation_design, .keywords = &.{ "zig", "code", "build.zig", "implement", "design" } },
    .{ .task = .code_repair, .keywords = &.{ "bug", "patch", "compile", "fix", "error" } },
    .{ .task = .legal_review, .keywords = &.{ "contract", "legal", "license", "compliance" } },
    .{ .task = .research_synthesis, .keywords = &.{ "paper", "research", "eval", "study" } },
    .{ .task = .benchmark_review, .keywords = &.{ "benchmark", "throughput", "latency", "perf" } },
    .{ .task = .project_recall, .keywords = &.{ "remember", "prior", "decision", "recall", "earlier" } },
};

/// Deterministically infer a `QueryPlan` from `query` by case-insensitive
/// substring matching (no allocation, no model call). Last-match-wins across the
/// keyword groups; `project_recall` additionally sets `exact_recall`.
pub fn infer(query: []const u8) QueryPlan {
    var plan = QueryPlan{ .query = query };

    for (KEYWORD_GROUPS) |group| {
        for (group.keywords) |kw| {
            if (containsIgnoreCase(query, kw)) {
                plan.task = group.task;
                break;
            }
        }
    }

    switch (plan.task) {
        .project_recall => {
            plan.exact_recall = true;
            // Recall favors provenance/exact wording over freshness.
            plan.recency_bias = 0.20;
        },
        .code_repair => {
            // Repair leans on the freshest structural state.
            plan.recency_bias = 0.60;
        },
        else => {},
    }

    return plan;
}

/// Case-insensitive substring search (ASCII). Local to keep `query_plan` free of
/// cross-feature imports for stub parity.
fn containsIgnoreCase(haystack: []const u8, needle: []const u8) bool {
    if (needle.len == 0) return true;
    if (needle.len > haystack.len) return false;
    var i: usize = 0;
    while (i + needle.len <= haystack.len) : (i += 1) {
        if (std.ascii.eqlIgnoreCase(haystack[i .. i + needle.len], needle)) return true;
    }
    return false;
}

test {
    std.testing.refAllDecls(@This());
}

test "infer defaults to general" {
    const plan = infer("what is the weather like today");
    try std.testing.expectEqual(TaskType.general, plan.task);
    try std.testing.expect(!plan.exact_recall);
}

test "infer detects implementation_design" {
    try std.testing.expectEqual(TaskType.implementation_design, infer("how do I write this in zig").task);
}

test "infer detects code_repair and raises recency bias" {
    const plan = infer("help me fix this compile bug");
    try std.testing.expectEqual(TaskType.code_repair, plan.task);
    try std.testing.expect(plan.recency_bias > 0.5);
}

test "infer detects legal_review" {
    try std.testing.expectEqual(TaskType.legal_review, infer("review this contract for compliance").task);
}

test "infer detects research_synthesis" {
    try std.testing.expectEqual(TaskType.research_synthesis, infer("summarize the research paper").task);
}

test "infer detects benchmark_review" {
    try std.testing.expectEqual(TaskType.benchmark_review, infer("what is the latency benchmark").task);
}

test "infer detects project_recall and enables exact_recall" {
    const plan = infer("remember the prior decision we made");
    try std.testing.expectEqual(TaskType.project_recall, plan.task);
    try std.testing.expect(plan.exact_recall);
    try std.testing.expect(plan.recency_bias < 0.40);
}

test "infer is case-insensitive" {
    try std.testing.expectEqual(TaskType.implementation_design, infer("Rewrite the CODE").task);
}

test "infer is last-match-wins: project_recall overrides earlier code match" {
    // contains both "code" (implementation_design) and "remember" (project_recall);
    // project_recall is later in precedence and wins.
    const plan = infer("remember the code decision");
    try std.testing.expectEqual(TaskType.project_recall, plan.task);
}

File: crates/abi-sea/src/query_plan.rs
//! `QueryPlan` / `TaskType` inference for SEA retrieval.
//!
//! Ported from `src/features/sea/query_plan.zig`. Deterministic keyword
//! heuristic — no model call.

/// Recognized retrieval intents.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub enum TaskType {
    /// No more-specific intent matched.
    #[default]
    General,
    /// Implementation / design work.
    ImplementationDesign,
    /// Bug fix / compile repair.
    CodeRepair,
    /// Legal / compliance review.
    LegalReview,
    /// Research synthesis.
    ResearchSynthesis,
    /// Project / decision recall (`project_recall`).
    ProjectRecall,
    /// Benchmark review.
    BenchmarkReview,
}

impl TaskType {
    /// `snake_case` tag name (matches Zig `@tagName`).
    #[must_use]
    pub const fn text(self) -> &'static str {
        match self {
            Self::General => "general",
            Self::ImplementationDesign => "implementation_design",
            Self::CodeRepair => "code_repair",
            Self::LegalReview => "legal_review",
            Self::ResearchSynthesis => "research_synthesis",
            Self::ProjectRecall => "project_recall",
            Self::BenchmarkReview => "benchmark_review",
        }
    }

    /// Stable ordinal used by the scorer's task-weight adjustments.
    #[must_use]
    pub const fn as_u8(self) -> u8 {
        match self {
            Self::General => 0,
            Self::ImplementationDesign => 1,
            Self::CodeRepair => 2,
            Self::LegalReview => 3,
            Self::ResearchSynthesis => 4,
            Self::ProjectRecall => 5,
            Self::BenchmarkReview => 6,
        }
    }
}

/// Inferred structured intent for a query.
#[derive(Debug, Clone, PartialEq)]
pub struct QueryPlan {
    /// Inferred task type.
    pub task: TaskType,
    /// Borrowed query text (never owned by the plan).
    pub query: String,
    /// Answers should be backed by selected evidence.
    pub require_grounding: bool,
    /// Prefer exact wording over fuzzy similarity.
    pub exact_recall: bool,
    /// Relative preference for fresher records, `[0,1]`.
    pub recency_bias: f32,
    /// Risk posture of the task, `[0,1]`.
    pub risk: f32,
}

/// Keyword groups in ascending precedence — later groups overwrite earlier
/// matches (last-match-wins).
const KEYWORD_GROUPS: &[(TaskType, &[&str])] = &[
    (
        TaskType::ImplementationDesign,
        &["zig", "code", "build.zig", "implement", "design"],
    ),
    (
        TaskType::CodeRepair,
        &["bug", "patch", "compile", "fix", "error"],
    ),
    (
        TaskType::LegalReview,
        &["contract", "legal", "license", "compliance"],
    ),
    (
        TaskType::ResearchSynthesis,
        &["paper", "research", "eval", "study"],
    ),
    (
        TaskType::BenchmarkReview,
        &["benchmark", "throughput", "latency", "perf"],
    ),
    (
        TaskType::ProjectRecall,
        &["remember", "prior", "decision", "recall", "earlier"],
    ),
];

/// Deterministically infer a [`QueryPlan`] from `query`.
#[must_use]
pub fn infer(query: &str) -> QueryPlan {
    let mut plan = QueryPlan {
        task: TaskType::General,
        query: query.to_string(),
        require_grounding: true,
        exact_recall: false,
        recency_bias: 0.40,
        risk: 0.50,
    };

    for &(task, keywords) in KEYWORD_GROUPS {
        for kw in keywords {
            if contains_ignore_case(query, kw) {
                plan.task = task;
                break;
            }
        }
    }

    match plan.task {
        TaskType::ProjectRecall => {
            plan.exact_recall = true;
            plan.recency_bias = 0.20;
        }
        TaskType::CodeRepair => {
            plan.recency_bias = 0.60;
        }
        _ => {}
    }

    plan
}

fn contains_ignore_case(haystack: &str, needle: &str) -> bool {
    if needle.is_empty() {
        return true;
    }
    if needle.len() > haystack.len() {
        return false;
    }
    let hay = haystack.as_bytes();
    let ned = needle.as_bytes();
    let mut i = 0;
    while i + ned.len() <= hay.len() {
        if eq_ignore_case_ascii(&hay[i..i + ned.len()], ned) {
            return true;
        }
        i += 1;
    }
    false
}

fn eq_ignore_case_ascii(a: &[u8], b: &[u8]) -> bool {
    a.eq_ignore_ascii_case(b)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_to_general() {
        let plan = infer("what is the weather like today");
        assert_eq!(plan.task, TaskType::General);
        assert!(!plan.exact_recall);
    }

    #[test]
    fn detects_each_task_class() {
        assert_eq!(
            infer("how do I write this in zig").task,
            TaskType::ImplementationDesign
        );
        assert_eq!(
            infer("help me fix this compile bug").task,
            TaskType::CodeRepair
        );
        assert!(infer("help me fix this compile bug").recency_bias > 0.5);
        assert_eq!(
            infer("review this contract for compliance").task,
            TaskType::LegalReview
        );
        assert_eq!(
            infer("summarize the research paper").task,
            TaskType::ResearchSynthesis
        );
        assert_eq!(
            infer("what is the latency benchmark").task,
            TaskType::BenchmarkReview
        );
        let recall = infer("remember the prior decision we made");
        assert_eq!(recall.task, TaskType::ProjectRecall);
        assert!(recall.exact_recall);
        assert!(recall.recency_bias < 0.40);
    }

    #[test]
    fn case_insensitive_and_last_match_wins() {
        assert_eq!(
            infer("Rewrite the CODE").task,
            TaskType::ImplementationDesign
        );
        // both "code" and "remember"; project_recall is later and wins
        assert_eq!(
            infer("remember the code decision").task,
            TaskType::ProjectRecall
        );
    }
}
